controllers/kvmcluster_controller.go

```diff
@@ -0,0 +1,159 @@
+	"sigs.k8s.io/cluster-api/util"
+	"sigs.k8s.io/cluster-api/util/annotations"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/giantswarm/cluster-api-provider-kvm/api/v1alpha4"
+	"github.com/giantswarm/cluster-api-provider-kvm/pkg/kvm/metrics"
+	"github.com/giantswarm/cluster-api-provider-kvm/pkg/kvm/scope"
+	"github.com/giantswarm/cluster-api-provider-kvm/pkg/kvm/services/namespace"
+	"github.com/pkg/errors"
+)
+
+const (
+	KVMClusterController = "kvmcluster"
+)
+
+// KVMClusterReconciler reconciles a KVMCluster object
+type KVMClusterReconciler struct {
+	client.Client
+	Recorder record.EventRecorder
+	Scheme   *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io.infrastructure.cluster.x-k8s.io,resources=kvmclusters,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io.infrastructure.cluster.x-k8s.io,resources=kvmclusters/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io.infrastructure.cluster.x-k8s.io,resources=kvmclusters/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the KVMCluster object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
+func (r *KVMClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reconcileError error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	defer metrics.CaptureMetrics()
+
+	// Fetch the KVMCluster instance
+	kvmCluster := &v1alpha4.KVMCluster{}
+	err := r.Get(ctx, req.NamespacedName, kvmCluster)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	defer func() {
+		if reconcileError != nil {
+			r.Recorder.Event(kvmCluster, "Warning", "failed-to-reconcile", reconcileError.Error())
+		}
+	}()
+
+	// Fetch the Cluster.
+	cluster, err := util.GetOwnerCluster(ctx, r.Client, kvmCluster.ObjectMeta)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if cluster == nil {
+		log.Info("Cluster Controller has not yet set OwnerRef")
+		return reconcile.Result{}, nil
+	}
+
+	if annotations.IsPaused(cluster, kvmCluster) {
+		log.Info("KVMCluster or linked Cluster is marked as paused. Won't reconcile")
+		return reconcile.Result{}, nil
+	}
+
+	log = log.WithValues("cluster", cluster.Name)
+
+	// Create the scope.
+	clusterScope, err := scope.NewClusterScope(scope.ClusterScopeParams{
```

Comment:
Scope is the object that contains all valuable information for this controller and is the thing that updates the objects when needed (in this case KVMCluster).

Reply:
This is a cool approach compared with putting random things into the context. Is this an upstream pattern or something you came up with?

Reply:
It's an upstream pattern. The scope contains things like kubernetes client and the objects you are reconciling over and ensures that they will get updated when they need to. Example cluster context in [CAPA](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/pkg/cloud/scope/cluster.go). It relies on a patch-helper which is an upstream util library (`sigs.k8s.io/cluster-api/util/patch`) that makes updating objects even easier.